[#236] 게시물: 누락된 기능 추가

jwonyLee opened the pull request:
## 개요

https://user-images.githubusercontent.com/15073405/142862772-e6c887d2-b01e-422b-82b5-1c70181ba24e.mp4

게시물 기능 추가

## 작업 사항

- 게시물 좋아요 기능 구현
- 게시물 삭제 기능 구현
  - 게시물을 삭제했을 때 더 이상 보여줄 게시물이 없다면 이전화면으로 돌아갑니다.
- 포토카드 화면으로 이동, 데이터 전달 구현

## 예외 사항

- 게시물 화면에서 삭제를 하고 난 후 이전 화면(홈, 모아보기)으로 돌아가면 반영되지 않고 있습니다. 해당 문제는 별도의 이슈로 만들어두었습니다. (미해결)
- PostTableCell.currentImagePage 값을 임시로 만들어서 전달하고 있습니다. 따라서 현재는 포토카드로 이동할 때 첫번째 이미지만 전달합니다.
- 게시물 수정 기능은 생각보다 로직이 길어질 거 같아 별도로 작업하려 합니다.

### Linked Issue

close #236

jwonyLee committed:
[#236] feat: 게시물 좋아요 기능 구현

jwonyLee committed:
[#236] feat: 게시물 삭제 기능 구현

jwonyLee committed:
[#238] feat: 포토카드 화면으로 이동, 데이터 전달 구현

jwonyLee committed:
Merge branch 'develop' of https://github.com/ThingLog/ThingLog into feature/236-post

gustn3965 commented on ThingLog/Coordinator/PhotoCardCoordinator.swift:
```diff
@@ -14,3 +14,11 @@ protocol PhotoCardCoordinatorProtocol: SystemSettingCoordinatorProtocol {
+        photoViewController.coordinator = self
+        navigationController.pushViewController(photoViewController, animated: true)
+    }
+}
```
`PhotoCard`에서는 하단 탭바를 숨겨야할 것 같습니당

gustn3965 commented on ThingLog/ViewController/Post/PostViewController.swift:
```diff
@@ -35,6 +36,8 @@ final class PostViewController: BaseViewController {
     override func viewDidLoad() {
         super.viewDidLoad()
+        viewModel.fetchedResultsController.delegate = self
+        viewModel.fetchedResultsControllerDelegate = self
```
궁금해서 그런데 이 코드가 필요한 이유가 있을까요?? `viewModel`의 `repository`에서 인자로 초기화하고 있는데, `delegate`를 언제 사용하는지도 궁금합니다.

gustn3965 commented on ThingLog/ViewController/Post/PostViewController.swift:
```diff
@@ -35,6 +36,8 @@ final class PostViewController: BaseViewController {
     // MARK: - Life Cycle
     override func viewDidLoad() {
         super.viewDidLoad()
+        viewModel.fetchedResultsController.delegate = self
```
* 게시물 화면에서 삭제를 하고 난 후 이전 화면(홈, 모아보기)으로 돌아가면 반영되지 않고 있습니다. 해당 문제는 별도의 이슈로 만들어두었습니다. (미해결)

위의 문제가 발생하는 원인이 이 코드입니다. 왜냐하면, `FetchedResultsController`가 새로운 객체가 아니라, 외부에서 주입받고 있는 형태인데, 이 객체가 외부에서도 여기서 구현한 `NSfetchedResultsDelegate`처럼, `delegate`를 이미 `self`로 주입받고있습니다. 그러므로 이 객체는 1-1로만 전달을 받을 수 있는데, 지금 여기서 하게 되면 외부에서 주입한 `self`가 풀리게 되어, 더이상 외부에서는 `delegate`를 받지 못합니다.   예로 들어, 홈화면에서 게시물을 선택한다음, 다시 홈화면으로 돌아와서 새로운 게시물을 작성하면 홈화면에서 업데이트가 되지 않습니다. 

사실 이 뷰컨트롤러가 `delegate`를 구현하지 않아도, 좋아요나, 수정은 이 작업은 어차피` fetchedResultsController`에 이미 반영이 되어 뷰에서도 적절히 보여줄 것으로 생각됩니다. 다만 삭제기능은 셀의 개수가 변경되는 것이니, 
`showRemovePostAlert()` 의 메소드안에서 삭제후, `tableView.reloadData()`를 호출하면 해결될 듯 합니다.
다만 지금` PostRepo`에서 `delete` 내부로직이 `perform`으로 되어있어, 셀의 개수가 맞지않다라고 에러가 뜨는데, 이는 `performAndWait`로 변경하거나, `completion`블록안에  `tableView.reloadData()` 를 호출하면 될 것 같습니다. 


추가적으로, `PostViewController`는 탭바를 숨기는게 더 좋을 것 같습니다. 그래야 `PostViewController`는 외부에서 발생하는 이벤트가 생길 수 없으므로 위에서 발생하는 Delegate문제를 더 줄여줄 것 같습니다.

gustn3965 commented on ThingLog/ViewModel/PostViewModel.swift:
```diff
@@ -21,10 +21,23 @@ protocol PostViewModelProtocol: AnyObject {
         self.fetchedResultsController = fetchedResultsController
         self.startIndexPath = startIndexPath
     }
+
+    func delete(_ entity: PostEntity) {
+        repository.delete([entity]) { result in
+            switch result {
+            case .success:
+                print("삭제 성공")
+            case .failure(let error):
+                fatalError("\(#function): \(error.localizedDescription)")
+            }
+        }
+    }
```
사실 이코드가 필요한지는 모르겠습니다.  `repository`가 `private`하여 감싸기 위한 목적도 아니고, 그렇다고 `repository`의 `update`코드는 없기 때문에 일관성이 없어보입니다.

gustn3965 reviewed (commented):
고생하셨습니다. 코멘트 확인해주세요 ~

jwonyLee committed:
[#236] ref: 피드백 반영
- 포토카드 화면으로 이동할 때 하단 탭바 숨김
- 게시물 화면으로 이동할 대 하단 탭바 숨김
- PostViewModel.delete 메서드 제거, showRemovePostAlert 에서 처리
- PostViewModel.fetchedResultsController.delegate 제거

jwonyLee commented:
피드백 주신 부분 반영해서 수정했습니다. 
- 포토카드 화면으로 이동할 때 하단 탭바 숨김
- 게시물 화면으로 이동할 대 하단 탭바 숨김
- PostViewModel.delete 메서드 제거, showRemovePostAlert 에서 처리
- PostViewModel.fetchedResultsController.delegate 제거

코멘트 남겨주신 대로 delegate를 지정해준 부분을 제거하니까 `게시물 화면에서 삭제를 하고 난 후 이전 화면(홈, 모아보기)으로 돌아가면 반영되지 않고 있습니다.` 문제가 발생하지 않는 걸로 확인했습니다. 제가 처음에 의도했던 것은 데이터가 삭제되었을 때, `표시할 수 있는 데이터가 없다면 이전 화면으로 돌아간다.`를 해결하기 위해 데이터가 변동되면 하기 코드를 실행하기 위해서였습니다.
```swift
// 표시할 수 있는 데이터가 없다면 이전 화면으로 이동한다.
if self?.viewModel.fetchedResultsController.fetchedObjects?.count ?? 0 == 0 {
    self?.coordinator?.back()
}
```
해당 부분은 게시물 삭제를 할 때 completion 블럭 내부에 호출하는 것으로 변경했습니다. 🙇‍♂️🙇‍♂️🙇‍♂️

gustn3965 reviewed (approved):
LGTM👍

jwonyLee committed:
Merge branch 'develop' into feature/236-post
